### PR TITLE
VB-4338 - Public booker sent cancel email

### PR DESCRIPTION
**Pull request by adaviesMOJ**
Initial commit. When a visit is cancelled, a visitor will be sent an email explaining why.

Next commit will rework to include multiple email templates depending on visit cancellation reason.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt [hunk 1]
```diff
@@ -64,25 +71,39 @@ class EmailSenderService(
...
+      ),
+    )
+
+    templateVars.putAll(getPrisonContactDetails(visit))
+
+    return SendEmailNotificationDto(templateName = EmailTemplateNames.VISIT_BOOKING, templateVars = templateVars)
+  }
+
+  private fun handleCancelledEvent(visit: VisitDto): SendEmailNotificationDto {
+    LOG.info("handleCancelEmail entered")
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt:94]: log message

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt [hunk 2]
```diff
@@ -36,12 +39,16 @@ class EmailSenderService(
...
 
-        else -> {
-          LOG.error("visit event type $visitEventType is unsupported")
+        CANCELLED -> {
+          handleCancelledEvent(visit)
+        }
+
+        UPDATED -> {
+          LOG.info("No email template for UPDATED event, skipping email notification for visit ${visit.reference}")
           return Unit
```
[the remark this answers is not in the document]
Comment (adaviesMOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt:53]: As mentioned on slack, you need to return from all when cases, as there is no UPDATED return we use Unit (void)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt (hunk 1 above)
Comment (adaviesMOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt:93]: Are you saying this should be

    LOG.info("handleCancelledEvent Entered")

?

Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt:93]: yes, similar to line 74

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt [hunk 3]
```diff
@@ -64,31 +73,77 @@ class EmailSenderService(
...
+        EmailTemplateNames.VISIT_CANCELLED_BY_PRISON
+      }
+
+      OutcomeStatus.VISITOR_CANCELLED -> {
+        EmailTemplateNames.VISIT_CANCELLED
+      }
+
+      else -> {
+        LOG.error("visit cancellation type $visitOutcome is unsupported")
+        throw ValidationException("visit cancellation type $visitOutcome is unsupported")
```
Comment (adaviesMOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt:118]: Spoke with Ed about other cancellation reasons. If public booker manages to send in a reason other than the ones supplied above, it's an error. So i've thrown a ValidationException in this case as it shouldn't have been received.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt [hunk 4]
```diff
@@ -64,31 +73,77 @@ class EmailSenderService(
...
+      }
+    }
+  }
+
+  private fun getCommonTemplateVars(visit: VisitDto): MutableMap<String, Any> {
```
[the remark this answers is not in the document]
Comment (adaviesMOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt:123]: Is there a difference / benefit in returning a normal map just to convert it?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/dto/visit/scheduler/VisitDto.kt [hunk 5]
```diff
@@ -28,4 +29,6 @@ data class VisitDto(
   val visitRestriction: VisitRestriction,
   @Schema(description = "List of visitors associated with the visit", required = true)
   val visitors: List<VisitorDto> = listOf(),
+  @Schema(description = "Outcome Status", example = "VISITOR_CANCELLED", required = false)
+  val outcomeStatus: OutcomeStatus? = null,
```
[the remark this answers is not in the document]
Comment (adaviesMOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/dto/visit/scheduler/VisitDto.kt:33]: okay no problem good suggestion, I'll get it updated!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (adaviesMOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt:123]: without mutable I can't add values to it after initialising the map, but I do this in a few places. Map doesn't have methods like put / putAll. I believe we need to keep it as a mutable map

File: src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt [hunk 6]
```diff
@@ -64,31 +73,77 @@ class EmailSenderService(
...
-    return SendEmailNotificationDto(templateName = templateName, templateVars = templateVars)
+    return templateVars
   }
 
-  private fun getPrisoner(visit: VisitDto): String {
+  private fun getPrisoner(visit: VisitDto): Map<String, Any> {
     return prisonerSearchService.getPrisoner(visit.prisonerId)?.let { prisoner ->
-      return "Your visit to see $prisoner"
-    } ?: "Your visit to the prison"
+      return mutableMapOf(
```
[the remark this answers is not in the document]
Comment (adaviesMOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/notificationsalertsvsip/service/EmailSenderService.kt:138]: same as above

**Review by dhirajshettymoj — APPROVED** [review]
LGTM

**Pull request merged**